### ISSUE TITLE
Horizon web request rate performance

### ISSUE THREAD

**Issue by oryband** [stellar/go#1446]
I was profiling where Horizon spends most of its CPU time within the app, other than having to wait for external services like History or Core database. I used [vegeta](https://github.com/tsenart/vegeta) (golang alternative to Apache Benchmark) to hit a single Horizon instance for 5m at 1000 HTTP GET req/s to the root endpoint `/`. These are the major CPU hoggers on that endpoint, by CPU time spent, ignoring the "must haves" which we can't do without (and thus no point in counting them as "performance hoggers") like Go garbage collection and time spent in `net/http` package: 

1. Building the HAL response, specifically the endpoint examples e.g. `/accounts/.../transactions/...`. This is mainly due to regex and string processing.
1. Gzip response encoding
1. Pretty printing the JSON response - indentation etc.
1. Metrics middleware
1. Logging middleware

The metrics and logging performance impact is very small compared to the others mentioned here.

By disabling the Gzip middleware, HAL response, and JSON pretty prettying from Horizon (commenting out lines or returning early from some functions) I managed to get Horizon request rate from around 500 req/s to 3000-4000 req/s.

I think that in some use cases Horizon can do without any HAL response and JSON pretty printing. In regards to Gzip response, I think it is crucial so let's drop that out of this discussion. For example, if my Horizon is only intended to be used as a client API without any human interaction, there's no need for HAL or pretty printing. I think we can add a configuration toggle for turning these off, but have them on by default to be backwards compatible.

Note: From what I understand this is not the primary bottleneck in Horizon, since the most popular use case would be other endpoints involving waiting for database queries to return. These in turn are limited by a maximum of a couple hundreds of concurrent connections to each database (according to some short googling on PostgreSQL limitations). Regardless, I think it make sense to propose the above.

Here is the stress test command I've used:
```bash
$ echo 'GET http://localhost:8000' | vegeta attack -rate=1000/s -duration=5m -insecure -output /opt/vegeta/results.bin
```
I've used [speescope](https://www.speedscope.app/) to visualize and analyze the pprof file into a flame graph. The command used was:
```bash
$ curl localhost:8000/debug/pprof/profile?seconds=310 > stress.5m.pprof
$ cat stress.5m.pprof | ./node_modules/speedscope/bin/cli.js -
```
The [pprof file can be downloaded here](https://github.com/stellar/go/files/3321339/stress.5m.pprof.zip), but here's a preview screenshot:

![Horizon speedscope pprof](https://user-images.githubusercontent.com/115387/60028396-e956de80-96a7-11e9-8352-e96e711990ae.png)

**Comment by oryband** [comment]
**EDIT:** Regarding the HAL response, I was specifically referring to the "custom" links under the `_links` section, specifically targeted for humans. The rest of the links can't be removed such as pagination links, etc.

**Comment by oryband** [comment]
Please ignore the commit message above, still working on the PR.

**Comment by bartekn** [comment]
@oryband thanks so much for the analysis. Sorry, I couldn't check it earlier. I took a closer look at the regexp in `PopulateTemplated`:
https://github.com/stellar/go/blob/69e628cf1cef67bbea02abc2e53ad3cddb60e4e8/support/render/hal/link.go#L14

It seems that we should be able to replace it with an alternative code that probably will be much faster like:
```go
l.Templated = strings.Contains(l.Href, "{")
```
I think we should try this first and compare execution times before using #1504.

**Comment by oryband** [comment]
Hey @bartekn . Thanks for the answer.

I implemented your proposal on top of my PR and here are the results:

To start, here's the pprof and flame graph results for using a recent version (https://github.com/stellar/go/commit/ff388105173f135586b35a18046e478209d18a87). I've submited 1000 req/s over 60s on the `/ root endpoint.

[ff388105.pprof.zip](https://github.com/stellar/go/files/3406720/ff388105.pprof.zip)

![ff388105](https://user-images.githubusercontent.com/115387/61462311-15017780-a97b-11e9-8d25-4ac8032b2a43.png)

As you can see regex code is taking around 10s in various "flames", all related to ``PopulateTemplated()`.

Next up is with the `strings.Contains()` implementation used instead of `regexp.Match()`. Human-designated HAL `"_links"` are turned on as well as indented "pretty" JSON response:

[hal-turned-on-use-strings-pkg-no-regex.pprof.zip](https://github.com/stellar/go/files/3406734/hal-turned-on-use-strings-pkg-no-regex.pprof.zip)

![hal-turned-on-use-strings-pkg-no-regex](https://user-images.githubusercontent.com/115387/61462515-6a3d8900-a97b-11e9-9953-bc3f095d736a.png)

The perfomance has improved considerably from ~10s to ~2s.

Next up is the implementation but without most of the HAL `"_links"` section turned off, except for pagination. JSON is still indented. There is not much to show in terms of flame graphs because that entire code section doesn't get executed on the `/` root endpoint.

[hal-turned-on.pprof.zip](https://github.com/stellar/go/files/3406735/hal-turned-on.pprof.zip)

Lastly, I've disabled indentation in JSON marshalling. The improvement here is minor <1s, but the reason I implemented it was not due to performance improvement, but simply because programmed clients using the API (SDKs etc.) don't need to parse newlines.

[hal-turned-off-and-unindented-json.pprof.zip](https://github.com/stellar/go/files/3406736/hal-turned-off-and-unindented-json.pprof.zip)

With "human" HAL `"_links"` disabled and using unindented JSON, I've managed to reach 2500 req/s on the same 2 vCPU EC2 instance used as in the first post on this issue. With the master version, using regex, human `"_links"` and indented JSON, I've managed to reach only 1500 req/s.

I think the scenarios I've described show there's value in having a toggle for disabling these kind of links and indented JSON, in addition to the `strings.Contains()` function. All HAL links and indented JSON will be enabled by default for backwards configuration. What do you think?

**Comment by ire-and-curses** [comment]
Hi @oryband, thanks for this! Replacing the regex looks like a great win that we should include.

Regarding the profiling results, have you tried any endpoints other than `/root`? I'm a little concerned that this work is micro-optimising performance on the `/root` endpoint, because that endpoint does very little except render. I am guessing that if you run your tests against an endpoint that does real work (e.g. `/transactions`) then the effect of these JSON and HAL optimisations on the overall speedup may be much smaller. If that was true then it probably wouldn't be worth adding the option to toggle.

Would you be willing to do a profile run against `/transactions` so we can get some sense of the real-world performance gain for a typical user?

**Comment by oryband** [comment]
Hey @ire-and-curses . I agree that besides avoiding regex, removing HAL population won't have a critical impact on performance. From what I've measured (using tools like [pganalyze](https://pganalyze.com/) the main bottleneck lies in database communication, specifically waiting on various SELECT, INSERT, and UPDATE queries.

The reason I suggested to add the abilitiy to toggle off HAL replies was primarily for usecases where Horizon is used for as an API gateway for mobile clients and not direct human interaction - HAL content is unnecessary, since programmable clients know which links to use according to the SDK used to develop them. For this usecase the only reason HAL is relevant is for pagination (and I've kept that in the PR).

Having said that, I also understand the reasoning to keep as few unnecessary code "bloat" in Horizon codebase as possible if it doesn't bring a lot of value.

Anyways, I can remove the HAL toggle and resubmit the PR if you'd like in hopes of getting this merged.

**Comment by oryband** [comment]
Hey @ire-and-curses and @bartekn . Please see the new PR #1582 .

In addition, please let me know if it is still relevant for you to implement the following:
1. Toggle for JSON marshaling without indentation (for SDK clients)
1. Toggle for HAL

I alerady have both implemented due to the discussions above and can cherry pick each of these features and submit in a new issue and PR if you desire.

**Comment by oryband** [comment]
ping
